loki/frontend/util.py

```diff
@@ -52,23 +52,23 @@ def __str__(self):
 REGEX = Frontend.REGEX
 
 
-def inline_comments(ir):
+class InlineCommentTransformer(Transformer):
     """
     Identify inline comments and merge them onto statements
     """
-    pairs = PatternFinder(pattern=(Assignment, Comment)).visit(ir)
-    pairs += PatternFinder(pattern=(VariableDeclaration, Comment)).visit(ir)
-    pairs += PatternFinder(pattern=(ProcedureDeclaration, Comment)).visit(ir)
-    mapper = {}
-    for pair in pairs:
-        # Comment is in-line and can be merged
-        # Note, we need to re-create the statement node
-        # so that Transformers don't throw away the changes.
-        if pair[0].source and pair[1].source:
-            if pair[1].source.lines[0] == pair[0].source.lines[1]:
-                mapper[pair[0]] = pair[0]._rebuild(comment=pair[1])
-                mapper[pair[1]] = None  # Mark for deletion
-    return NestedTransformer(mapper, invalidate_source=False).visit(ir)
+
+    def visit_tuple(self, o, **kwargs):
+        pairs = PatternFinder(pattern=(Assignment, Comment)).visit(o)
+        pairs += PatternFinder(pattern=(VariableDeclaration, Comment)).visit(o)
+        pairs += PatternFinder(pattern=(ProcedureDeclaration, Comment)).visit(o)
+
+        for pair in pairs:
+            # Comment is in-line and can be merged
+            if pair[0].source and pair[1].source:
+                if pair[1].source.lines[0] == pair[0].source.lines[1]:
+                    new = pair[0]._rebuild(comment=pair[1])
+                    o = replace_windowed(o, pair, new)
+        return o
```

Comment:
Slightly confused by the implementation:

Undoubtedly, the performance gain comes here from not using `NestedTransformer` and instead doing in-place updates for the found nodes. However, I don't understand why this is implemented as a `Transformer`, and not convinced it does always behave as expected?

It shortcuts the recursion in the `Transformer` upon the first tuple - which will in the frontend use-case usually be the `Section.body` for something like spec, body in a subroutine. But the PatternFinder is then the only place to actually recurse into the IR. Does the replacement really work as expected for inline comments further nested into the IR than the outermost tuple?

Unless we're explicitly testing for this, it wouldn't upset anything because it just means the inline comments continue to live as standalone comments. So, I'm wondering whether the effect is merely that we don't convert the comments to inline comments at all anymore?

Reply:
Wow, very good spot! You are totally right that 
a) this isn't tested properly!
b) We recurse in `PatternFinder`, which we don't need anymore, but not in `visit_tuple`

I've tried to address both in two commits I just pushed, and with those the behaviour seems correct, and performance benefits stay as before. Please have another look (once CI agrees)

Reply:
Funny, this also fixed the `test_ir_graph`, so that I could remove the commit that adjusts the node/edge counting test (hence recent force-push).